pkg/splunk/enterprise/util.go

```diff
@@ -34,15 +37,244 @@ import (
+			existingLatestVersionedSecret = secret
+		}
+		secretFound = true
+	}
+
+	return &existingLatestVersionedSecret, existingLatestVersion, secretFound
+}
+
+// GetLatestVersionedSecret is used to create/retreive latest versionedSecretIdentifier based secret, cr is optional for owner references(pass nil if not required)
+func GetLatestVersionedSecret(c splcommon.ControllerClient, cr splcommon.MetaObject, namespace string, versionedSecretIdentifier string) (*corev1.Secret, error) {
+	var latestVersionedSecret *corev1.Secret
+	var err error
+	var secretFound bool
+
+	// Retreive namespaced scoped secret data in splunk readable format
+	splunkReadableData, err := GetSplunkReadableNamespaceScopedSecretData(c, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the latest versionedSecretIdentifier based secret, if atleast one exists
+	existingLatestVersionedSecret, existingLatestVersion, secretFound := GetExistingLatestVersionedSecret(c, namespace, versionedSecretIdentifier)
+
+	// Check if there is atleast one versionedSecretIdentifier based secret
+	if secretFound == false {
+		// No secret based on versionedSecretIdentifier, create one with version v1
+		latestVersionedSecret, err = ApplySplunkSecret(c, cr, splunkReadableData, GetVersionedSecretName(versionedSecretIdentifier, firstVersion), namespace)
+	} else {
+		if existingLatestVersionedSecret != nil {
+			// Check if contents of latest versionedSecretIdentifier based secret is different from that of namespace scoped secrets object
+			if !reflect.DeepEqual(splunkReadableData, existingLatestVersionedSecret.Data) {
+				// Different, create a newer version versionedSecretIdentifier based secret
+				latestVersionedSecret, err = ApplySplunkSecret(c, cr, splunkReadableData, GetVersionedSecretName(versionedSecretIdentifier, strconv.Itoa(existingLatestVersion+1)), namespace)
+				return latestVersionedSecret, err
+			}
+			// Latest versionedSecretIdentifier based secret is the the existing latest versionedSecretIdentifier based secret
+			latestVersionedSecret = existingLatestVersionedSecret
+		}
+	}
+
+	return latestVersionedSecret, nil
+}
+
+// GetSplunkReadableNamespaceScopedSecretData retreives the namespace scoped secret's data and converts it into Splunk readable format if possible
+func GetSplunkReadableNamespaceScopedSecretData(c splcommon.ControllerClient, namespace string) (map[string][]byte, error) {
+	// Get namespace scoped secret
+	namespaceScopedSecret, err := GetNamespaceScopedSecret(c, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	secretTokenTypes := []string{"hec_token", "password", "pass4SymmKey", "idxc_secret", "shc_secret"}
+	for _, tokenType := range secretTokenTypes {
+		if _, ok := namespaceScopedSecret.Data[tokenType]; !ok {
+			return nil, fmt.Errorf(missingTokenError, tokenType)
+		}
+	}
+
+	// Prepare versionedSecretIdentifier based secret data
+	return map[string][]byte{
+		"hec_token":    namespaceScopedSecret.Data["hec_token"],
```

Comment:
Here I have used a common API for individual token types but for default.yml used the raw values. I could use a map of map i.e map[token_type] = token_type to ensure I can access all token_type strings at one go but it might be an overkill.